### PR TITLE
Add token to the request object through handleToken method

### PR DESCRIPTION
**Pull request by dolfelt**
@shadowhand @elazar

### CODE REVIEW

**Comment by elazar** [comment]
Looks like we've got test failures. Related?

File: src/AuthHandler.php [hunk 1]
```diff
@@ -63,17 +63,24 @@ public function __invoke(
...
                 throw new UnauthorizedException;
             }
+
+            $request = $this->handleToken($authToken, $request);
         }
 
         return $next($request, $response);
     }
 
+    public function handleToken(Token $token, ServerRequestInterface $request)
```
Comment (shadowhand) [inline src/AuthHandler.php]: This only works for tokens, not for credentials. https://github.com/sparkphp/auth/blob/master/src/Credentials.php

Comment (shadowhand) [inline src/AuthHandler.php]: You'll need separate methods for each.

And this method should be `private`

Comment (elazar) [inline src/AuthHandler.php]: Eh? Both validate methods return a Token object.

Comment (shadowhand) [inline src/AuthHandler.php]: Oh whoops. Right you are.

**Comment by dolfelt** [comment]
Fixed the tests and updated to be `private`.

File: src/AuthHandler.php [hunk 2]
```diff
@@ -63,18 +63,35 @@ public function __invoke(
...
         }
 
         return $next($request, $response);
     }
 
     /**
+     * Handle the Token and put it in the request
+     *
+     * @param Token $token
+     * @param ServerRequestInterface $request
```
Comment (elazar) [inline src/AuthHandler.php]: Parameter order is backwards.

File: src/AuthHandler.php [hunk 3]
```diff
@@ -63,18 +63,35 @@ public function __invoke(
...
+     * @return ServerRequestInterface
+     */
+    private function handleToken(ServerRequestInterface $request, Token $token = null)
```
Comment (shadowhand) [inline src/AuthHandler.php]: In what scenario would the `$token = null`? Wouldn't that mean we failed to validate and should be throwing an exception?

Comment (dolfelt) [inline src/AuthHandler.php]: Not sure, but the tests support it. Needed to add that to make the existing tests pass.

Comment (shadowhand) [inline src/AuthHandler.php]: Okay, so why not fix the calling code?

File: src/AuthHandler.php [hunk 4]
```diff
@@ -63,18 +63,35 @@ public function __invoke(
...
-                $this->adapter->validateToken($token);
+                $authToken = $this->adapter->validateToken($token);
             } elseif ($credentials = $this->credentials->getCredentials($request)) {
-                $this->adapter->validateCredentials($credentials);
+                $authToken = $this->adapter->validateCredentials($credentials);
             } else {
                 throw new UnauthorizedException;
             }
+
+            $request = $this->handleToken($request, $authToken);
```
Comment (shadowhand) [inline src/AuthHandler.php]: Wrap this with `if ($authToken)` instead.

Comment (dolfelt) [inline src/AuthHandler.php]: To be more flexible / extensible.

File: src/AuthHandler.php (hunk 3 above)
Comment (elazar) [inline src/AuthHandler.php]: Token definitely should not be null. Note @return tags in AdapterInterface. There must be something wrong with the tests.

File: src/AuthHandler.php (hunk 4 above)
Comment (elazar) [inline src/AuthHandler.php]: Shouldn't be necessary since the return value of either validate method shouldn't be null.

Comment (shadowhand) [inline src/AuthHandler.php]: But the method is private, it can't be extended.

Comment (elazar) [inline src/AuthHandler.php]: Not seeing why we'd need it to be?

Comment (shadowhand) [inline src/AuthHandler.php]: Comment is outdated. :) It is fixed elsewhere by ensuring the token exists.

File: tests/AuthHandlerTest.php [hunk 5]
```diff
@@ -18,6 +18,10 @@
 
 class AuthHandlerTest extends TestCase
 {
+
+    const AUTH_TOKEN_NAME = 'spark/auth:token';
+    const AUTH_TOKEN_STRING = 'token';
```
Comment (shadowhand) [inline tests/AuthHandlerTest.php]: :+1:

File: src/AuthHandler.php [hunk 6]
```diff
@@ -63,18 +63,32 @@ public function __invoke(
...
     }
 
     /**
+     * Handle the Token and put it in the request
+     *
+     * @param ServerRequestInterface $request
+     * @param Token $token
+     * @return ServerRequestInterface
+     */
+    protected function handleToken(ServerRequestInterface $request, Token $token)
```
Comment (dolfelt) [inline src/AuthHandler.php]: Changed this to `protected` because I really want the ability to overwrite it.

**Comment by dolfelt** [comment]
What do you think of this @elazar @ameech @shadowhand?

**Comment by shadowhand** [comment]
I think we're good now. :)

**Pull request merged**